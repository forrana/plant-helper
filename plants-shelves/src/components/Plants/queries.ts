import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
    symbol {
      userWideId
    }
    room {
      id,
      roomName,
      colorBackground
    }
  }
}
`;

const GET_ALL_ROOMS = gql`
  query {
    rooms {
      id, roomName, colorBackground
    }
  }
`;
const ROOMS_BY_NAME_FRAGMENT = gql`
  query RoomsByNameFragment($nameFragment: String!) {
    roomsByNameFragment(nameFragment: $nameFragment) {
      id, roomName, colorBackground
    }
  }
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $groupName: String, $colorBackground: String) {
      createPlant(plantName: $plantName, scientificName: $scientificName, daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
        daysBetweenWateringDormant: $daysBetweenWateringDormant, groupName: $groupName, colorBackground: $colorBackground) {
        plant {
          id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
          symbol {
            userWideId
          }
          room {
            id,
            roomName,
            colorBackground
          }
        }
      }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const POSTPONE_WATERING = gql`
  mutation PostponeWatering($plantId: ID!, $postponeDays: Int!) {
    postponeWatering(plantId: $plantId, days: $postponeDays) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_PLANT = gql`
  mutation UpdatePlant($plantId: ID!, $plantName: String!,
    $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $postponeDays: Int!, $groupName: String, $colorBackground: String) {
    updatePlant(plantId: $plantId, plantName: $plantName, scientificName: $scientificName,
      daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
      daysBetweenWateringDormant: $daysBetweenWateringDormant,
      postponeDays: $postponeDays,
      groupName: $groupName,
      colorBackground: $colorBackground) {
      plant {
        id, name, scientificName, daysUntilNextWatering,
        daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant,
        daysPostpone,
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_ROOM = gql`
  mutation UpdateRoom($roomId: ID!, $roomName: String!) {
    updateRoom(roomId: $roomId, roomName: $roomName) {
      ok
      room {
        id,
        roomName,
        colorBackground
      }
    }
  }
`

const CREATE_ROOM = gql`
  mutation CreateRoom($plant1Id: ID!, $plant2Id: ID!) {
    createRoom(plant1Id: $plant1Id, plant2Id: $plant2Id) {
      ok
    }
  }
`;

const DELETE_PLANT = gql`
  mutation DeletePlant($plantId: ID!) {
    deletePlant(plantId: $plantId) {
      ok
    }
  }
`;

const GET_SUBSCRIPTION = gql`
  query {
    subscription
  }
`

const CREATE_SUBSCRIPTION = gql`
  mutation CreateSubscription($endpoint: String!, $p256dh: String!, $auth: String!, $permissionGiven: Boolean!) {
    createSubscription(endpoint: $endpoint, p256dh: $p256dh, auth: $auth, permissionGiven: $permissionGiven) {
      ok
    }
  }
`

export {
  GET_ALL_PLANTS,
  GET_ALL_ROOMS,
  ADD_PLANT,
  WATER_PLANT,
  UPDATE_PLANT,
  DELETE_PLANT,
  CREATE_SUBSCRIPTION,
  GET_SUBSCRIPTION,
  POSTPONE_WATERING,
  CREATE_ROOM,
  ROOMS_BY_NAME_FRAGMENT,
  UPDATE_ROOM
}
