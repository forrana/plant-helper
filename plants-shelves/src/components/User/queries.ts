import { gql } from '@apollo/client';

const CREATE_USER = gql`
  mutation Register($email: String!, $username: String!, $password1: String!, $password2: String!) {
    register(email: $email, username: $username, password1: $password1, password2: $password2) {
      success,
      errors
    }
  }
`;

const LOG_IN = gql`
  mutation TokenAuth($username: String!, $password: String!) {
    tokenAuth(username: $username, password: $password) {
      success,
      errors,
      unarchiving,
      token,
      refreshToken,
      unarchiving,
      user {
        id,
        username,
      }
    }
  }
`;


const REFRESH_TOKEN = gql`
  mutation RefreshToken($refreshToken: String!) {
    refreshToken(
      refreshToken: $refreshToken
    ) {
      success,
      errors,
      payload,
      token,
      refreshToken,
    }
  }
`

const UPSERT_USER_SETTINGS = gql`
  mutation UpsertUserSettings($startTime: String!, $endTime: String!, $timezone: String) {
    upsertUserSettings(startTime: $startTime, endTime: $endTime, timezone: $timezone) {
      ok
    }
  }
`

const GET_USER_SETTINGS = gql`
  query {
    userSettings {
      notificationsStartTime, notificationsEndTime, timezone
    }
  }
`;

export { CREATE_USER, LOG_IN, REFRESH_TOKEN, UPSERT_USER_SETTINGS, GET_USER_SETTINGS }
