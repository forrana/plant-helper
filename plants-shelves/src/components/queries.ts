import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
  }
}
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!) {
    createPlant(plantName: $plantName, scientificName: $scientificName) {
      plant {
        name,
        scientificName
      }
    }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
      }
    }
  }
`;


export { GET_ALL_PLANTS, ADD_PLANT, WATER_PLANT }
