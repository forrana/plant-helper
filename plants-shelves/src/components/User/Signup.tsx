import React, { useState } from 'react';
import { Button, Form, FormGroup, Label, Input, Spinner, FormFeedback } from 'reactstrap';
import { useMutation } from '@apollo/client';

import { CREATE_USER } from './queries'
import styles from "./Login.module.css"
import { SignupErrors, Error } from './models';

function Signup() {
  const [username, setLogin] = useState("");
  const [password1, setPassword] = useState("");
  const [password2, setPasswordConfirmation] = useState("");
  const [email, setEmail] = useState("");
  const [signupErrors, setSignupErrors] = useState<SignupErrors>({});

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const name = event.target.name;
      resetFieldErrors(name);
      switch(name) {
        case "username": setLogin(event.target.value); break;
        case "password1": setPassword(event.target.value); break;
        case "password2": setPasswordConfirmation(event.target.value); break;
        case "email": setEmail(event.target.value); break;
      }
  }

  const [createUser, { loading, error }] = useMutation(CREATE_USER, {
    onCompleted: (data: any) => {
      const errors: SignupErrors = data?.register?.errors;
      if(errors) {
        setSignupErrors(errors);
      } else {
        setSignupErrors({});
        setLogin("");
        setPassword("");
        setPasswordConfirmation("");
        setEmail("");
      }
    },
    onError: (e) => console.error('Signup error:', e)
  });

  const getFieldErrors = (field: string): Array<Error> => {
    let result: Array<Error> = [];
    if(signupErrors[field]) {
      result = [...result, ...signupErrors[field]];
    }
    if(signupErrors['nonFieldErrors']) {
      result = [...result, ...signupErrors['nonFieldErrors']];
    }
    return result;
  }

  const resetFieldErrors = (field: string) => {
    setSignupErrors({...signupErrors, [field]: []})
  }

  const isFieldHasErrors = (field: string): boolean => getFieldErrors(field).length > 0;

  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    createUser({ variables: { username, email, password1, password2 } })
  }

  if (loading) return  <Spinner color="primary" />

  if (error)  return  <p>Error :( {error.message}</p>;

  return (
    <main>
      <Form
        onSubmit={handleFormSubmit}
        autoComplete="off"
        className={styles.container}
      >
        <FormGroup>
          <Label for="login">Login:</Label>
          <Input type="text" name="username" id="login" placeholder="Enter login"
            autoComplete="off"
            value={username}
            onChange={handleInputChange}
            invalid={isFieldHasErrors("username")}
            required
          />
          {
            getFieldErrors("username").map((error, index) =>
              <FormFeedback key={index}>{ error.message }</FormFeedback>
            )
          }
        </FormGroup>
        <FormGroup>
          <Label for="email">Email:</Label>
          <Input type="email" name="email" id="email" placeholder="Enter email"
            autoComplete="off"
            value={email}
            onChange={handleInputChange}
            invalid={isFieldHasErrors("email")}
            required
          />
          {
            getFieldErrors("email").map((error, index) =>
              <FormFeedback key={index}>{ error.message }</FormFeedback>
            )
          }
        </FormGroup>
        <FormGroup>
          <Label for="password">Password:</Label>
          <Input type="password" name="password1" id="password" placeholder="Enter password"
            autoComplete="off"
            value={password1}
            onChange={handleInputChange}
            invalid={isFieldHasErrors("password1")}
            required
          />
          {
            getFieldErrors("password1").map((error, index) =>
              <FormFeedback key={index}>{ error.message }</FormFeedback>
            )
          }
        </FormGroup>
        <FormGroup>
          <Label for="password-confirmation">Password confirmation:</Label>
          <Input type="password" name="password2" id="password-confirmation" placeholder="Re-enter password"
            autoComplete="off"
            value={password2}
            onChange={handleInputChange}
            invalid={isFieldHasErrors("password2")}
            required
          />
          {
            getFieldErrors("password2").map((error, index) =>
              <FormFeedback key={index}>{ error.message }</FormFeedback>
            )
          }
        </FormGroup>
        <section className={styles.controls}>
          <Button type="submit">Create</Button>
        </section>
      </Form>
    </main>
  )
}

export default Signup
