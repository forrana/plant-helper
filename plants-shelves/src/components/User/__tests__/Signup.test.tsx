import React from 'react';
import { act, fireEvent, render, screen, Screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import { MemoryRouter } from 'react-router-dom';
import Signup from '../Signup';
import { CREATE_USER } from '../queries';

const user = {
  username: "username",
  password1: "password" ,
  password2: "password",
  email: "test@email.com"
 };

const error = {
  message: "Please, enter valid credentials.",
  code: "invalid_credentials"
}

const genericErrorMessage = "Something went wrong"

const mocksWithUser: any = [
  {
    request: {
      query: CREATE_USER,
      variables: {
        ...user
      }
    },
    result: {
      data: {
        register: {
          success: true,
          errors: []
        }
      }
    },
  },
];

const mocksWithErrors: any = [
  {
    request: {
      query: CREATE_USER,
      variables: {
        ...user
      }
    },
    result:{
      data: {
        register: {
          errors: {
            nonFieldErrors:[
              error
            ]
          }
        }
      }
    }
  },
  {
    request: {
      query: CREATE_USER,
      variables: {
        ...user
      }
    },
    error:{
      message: genericErrorMessage
    }
  }
];

test('Match snapshot', async () => {
  const {container} = render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithUser} addTypename={false}>
        <Signup />
      </MockedProvider>
    </MemoryRouter>,
  );

  expect(container).toMatchSnapshot()
});

test('Allow to signup', async () => {
 const { container } = render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithUser} addTypename={false}>
      <Signup />
      </MockedProvider>
    </MemoryRouter>,
  );

  await act(async () => {
    fireEvent.change(
      screen.getByLabelText("Login:"),
      { target: { value: user.username } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByLabelText("Email:"),
      { target: { value: user.email } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByPlaceholderText("Enter password:"),
      { target: { value: user.password1 } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByPlaceholderText("Re-enter password:"),
      { target: { value: user.password1 } }
    )
  });

  await act(async () => {
    fireEvent.click(
      screen.getByTestId("login-submit-button")
    )
    await new Promise(resolve => setTimeout(resolve, 0));
  });

  expect(screen.getByLabelText("Password:")).not.toHaveValue(user.password1)
})

// test('Don\'t submit with empty fields', async () => {
//   const { container } = render(
//      <MemoryRouter>
//        <MockedProvider mocks={mocksWithUser} addTypename={false}>
//        <Signup />
//        </MockedProvider>
//      </MemoryRouter>,
//    );

//    await act(async () => {
//      fireEvent.change(
//        screen.getByLabelText("Login:"),
//        { target: { value: null } }
//      )
//    });

//    await act(async () => {
//      fireEvent.change(
//        screen.getByPlaceholderText("Enter password:"),
//        { target: { value: user.password1 } }
//      )
//    });

//    await act(async () => {
//      fireEvent.click(
//        screen.getByTestId("login-submit-button")
//      )
//      await new Promise(resolve => setTimeout(resolve, 0));
//    });

//    expect(screen.getByLabelText("Password:")).toHaveValue(user.password)
//  })

// test('Show login errors', async () => {
//   const { container } = render(
//      <MemoryRouter>
//        <MockedProvider mocks={mocksWithErrors} addTypename={false}>
//        <Signup />
//        </MockedProvider>
//      </MemoryRouter>,
//    );

//    await act(async () => {
//      fireEvent.change(
//        screen.getByLabelText("Login:"),
//        { target: { value: user.username } }
//      )
//    });

//    await act(async () => {
//      fireEvent.change(
//        screen.getByLabelText("Password:"),
//        { target: { value: user.password } }
//      )
//    });

//    await act(async () => {
//      fireEvent.click(
//        screen.getByTestId("login-submit-button")
//      )
//      await new Promise(resolve => setTimeout(resolve, 0));
//    });

//    expect(screen.getByTestId("login-error-message")).toBeVisible()
//  })

//  test('Show error if there is no token returned', async () => {
//   const { container } = render(
//      <MemoryRouter>
//        <MockedProvider mocks={mocksWithEmptyToken} addTypename={false}>
//        <Signup />
//        </MockedProvider>
//      </MemoryRouter>,
//    );

//    await act(async () => {
//      fireEvent.change(
//        screen.getByLabelText("Login:"),
//        { target: { value: user.username } }
//      )
//    });

//    await act(async () => {
//      fireEvent.change(
//        screen.getByLabelText("Password:"),
//        { target: { value: user.password } }
//      )
//    });

//    await act(async () => {
//      fireEvent.click(
//        screen.getByTestId("login-submit-button")
//      )
//      await new Promise(resolve => setTimeout(resolve, 0));
//    });

//    expect(screen.getByTestId("login-error-message")).toBeVisible()
//  })

//  test('Show general error on server error', async () => {
//   const { container } = render(
//      <MemoryRouter>
//        <MockedProvider mocks={mocksWithErrors} addTypename={false}>
//        <Signup />
//        </MockedProvider>
//      </MemoryRouter>,
//    );

//    await act(async () => {
//      fireEvent.change(
//        screen.getByLabelText("Login:"),
//        { target: { value: "error" } }
//      )
//    });

//    await act(async () => {
//      fireEvent.change(
//        screen.getByLabelText("Password:"),
//        { target: { value: user.password } }
//      )
//    });

//    await act(async () => {
//      fireEvent.click(
//        screen.getByTestId("login-submit-button")
//      )
//      await new Promise(resolve => setTimeout(resolve, 0));
//    });

//    expect(screen.getByText(new RegExp(genericErrorMessage, "i"))).toBeVisible()
//  })