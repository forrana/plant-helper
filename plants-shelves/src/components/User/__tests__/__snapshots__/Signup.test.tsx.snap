// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
<div>
  <main>
    <form
      autocomplete="off"
      class="container"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="login"
        >
          Login:
        </label>
        <input
          aria-invalid="false"
          autocomplete="off"
          class="form-control"
          id="login"
          name="username"
          placeholder="Enter login"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class=""
          for="email"
        >
          Email:
        </label>
        <input
          aria-invalid="false"
          autocomplete="off"
          class="form-control"
          id="email"
          name="email"
          placeholder="Enter email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class=""
          for="password"
        >
          Password:
        </label>
        <input
          aria-invalid="false"
          autocomplete="off"
          class="form-control"
          id="password"
          name="password1"
          placeholder="Enter password"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class=""
          for="password-confirmation"
        >
          Password confirmation:
        </label>
        <input
          aria-invalid="false"
          autocomplete="off"
          class="form-control"
          id="password-confirmation"
          name="password2"
          placeholder="Re-enter password"
          required=""
          type="password"
          value=""
        />
      </div>
      <section
        class="controls"
      >
        <button
          class="btn btn-secondary"
          data-testid="signup-submit-button"
          type="submit"
        >
          Create
        </button>
        <button
          class="btn btn-link"
        >
          <a
            href="/login"
          >
            Login
          </a>
        </button>
      </section>
    </form>
  </main>
</div>
`;
